File: packages/protocols/x402-libertai/examples/agentkit/typescript/src/agent.ts
import { AgentKit, erc20ActionProvider, walletActionProvider } from "@coinbase/agentkit";
import {
  actionsToTools,
  createAgentWallet,
  createAlephActionProvider,
  createLLMClient,
} from "@libertai/agentkit-plugin";
import type { ChatCompletionMessageParam } from "openai/resources/chat/completions.mjs";
import { config } from "./config.js";

const SYSTEM_PROMPT = `You are an autonomous AI agent on Base blockchain.
You have a wallet with ETH and USDC. You pay for compute via Aleph credits.

Each cycle:
1. Check your balances (ETH, USDC) and credit info
2. If credits are running low, buy more credits
3. If ETH is too low, swap some USDC for ETH
4. Report your status

Be concise. Only take action when needed.`;

export async function startAgent() {
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  const wallet = await createAgentWallet(config.privateKey, config.chain as any, config.rpcUrl);
  console.log(`Wallet: ${wallet.account.address}`);

  const agentkit = await AgentKit.from({
    walletProvider: wallet.provider,
    actionProviders: [
      walletActionProvider(),
      erc20ActionProvider(),
      createAlephActionProvider(config.privateKey),
    ],
  });

  const { tools, executeTool } = actionsToTools(agentkit.getActions());
  const openai = createLLMClient(config.privateKey);

  console.log(`Model: ${config.model}`);
  console.log(`Cycle interval: ${config.cycleIntervalMs}ms`);
  console.log(`Tools: ${tools.map((t) => t.function.name).join(", ")}`);
  console.log("---");

  let cycle = 0;
  while (true) {
    cycle++;
    console.log(`\n=== Cycle ${cycle} ===`);

    const messages: ChatCompletionMessageParam[] = [
      { role: "system", content: SYSTEM_PROMPT },
      { role: "user", content: `Cycle ${cycle}: Check status and take any necessary actions.` },
    ];

    try {
      let response = await openai.chat.completions.create({
        model: config.model,
        messages,
        tools,
      });

      while (response.choices[0]?.finish_reason === "tool_calls") {
        const assistantMessage = response.choices[0].message;
        messages.push(assistantMessage);

        for (const toolCall of assistantMessage.tool_calls ?? []) {
          const name = toolCall.function.name;
          const args = toolCall.function.arguments;
          console.log(`  -> ${name}(${args})`);

          const result = await executeTool(name, args);

          messages.push({
            role: "tool",
            tool_call_id: toolCall.id,
            content: result,
          });
        }

        response = await openai.chat.completions.create({
          model: config.model,
          messages,
          tools,
        });
      }

      const text = response.choices[0]?.message?.content;
      if (text) console.log(`Agent: ${text}`);
    } catch (err) {
      console.error("Cycle error:", err);
    }

    console.log(`Waiting ${config.cycleIntervalMs}ms...`);
    await new Promise((r) => setTimeout(r, config.cycleIntervalMs));
  }
}

File: packages/protocols/x402-libertai/examples/agentkit/python/agent.py
from __future__ import annotations

import asyncio

from coinbase_agentkit import AgentKit, AgentKitConfig
from coinbase_agentkit.action_providers.erc20.erc20_action_provider import (
    ERC20ActionProvider,
)
from coinbase_agentkit.action_providers.wallet.wallet_action_provider import (
    WalletActionProvider,
)

from libertai_agentkit_plugin import (
    actions_to_tools,
    create_agent_wallet,
    create_llm_client,
)
from libertai_agentkit_plugin.actions.aleph import AlephActionProvider

from config import CYCLE_INTERVAL_S, MODEL, PRIVATE_KEY, RPC_URL

SYSTEM_PROMPT = """You are an autonomous AI agent on Base blockchain.
You have a wallet with ETH and USDC. You pay for compute via Aleph credits.

Each cycle:
1. Check your balances (ETH, USDC) and credit info
2. If credits are running low, buy more credits
3. If ETH is too low, swap some USDC for ETH
4. Report your status

Be concise. Only take action when needed."""


async def start_agent() -> None:
    wallet_provider = create_agent_wallet(PRIVATE_KEY, rpc_url=RPC_URL)
    address = wallet_provider.get_address()
    print(f"Wallet: {address}")

    agentkit = AgentKit(AgentKitConfig(
        wallet_provider=wallet_provider,
        action_providers=[
            WalletActionProvider(),
            ERC20ActionProvider(),
            AlephActionProvider(PRIVATE_KEY),
        ],
    ))

    tools, execute_tool = actions_to_tools(agentkit.get_actions())
    openai = create_llm_client(PRIVATE_KEY)

    print(f"Model: {MODEL}")
    print(f"Cycle interval: {CYCLE_INTERVAL_S}s")
    print(f"Tools: {', '.join(t['function']['name'] for t in tools)}")
    print("---")

    cycle = 0
    while True:
        cycle += 1
        print(f"\n=== Cycle {cycle} ===")

        messages = [
            {"role": "system", "content": SYSTEM_PROMPT},
            {"role": "user", "content": f"Cycle {cycle}: Check status and take any necessary actions."},
        ]

        try:
            response = await openai.chat.completions.create(
                model=MODEL,
                messages=messages,
                tools=tools,
            )

            while response.choices[0].finish_reason == "tool_calls":
                assistant_msg = response.choices[0].message
                messages.append(assistant_msg)

                for tool_call in assistant_msg.tool_calls or []:
                    name = tool_call.function.name
                    args = tool_call.function.arguments
                    print(f"  -> {name}({args})")

                    result = await execute_tool(name, args)
                    messages.append({
                        "role": "tool",
                        "tool_call_id": tool_call.id,
                        "content": result,
                    })

                response = await openai.chat.completions.create(
                    model=MODEL,
                    messages=messages,
                    tools=tools,
                )

            text = response.choices[0].message.content
            if text:
                print(f"Agent: {text}")
        except Exception as e:
            print(f"Cycle error: {e}")

        print(f"Waiting {CYCLE_INTERVAL_S}s...")
        await asyncio.sleep(CYCLE_INTERVAL_S)


if __name__ == "__main__":
    asyncio.run(start_agent())
